Add support for Minecraft 1.20.3

Hello,

Minecraft 1.20.3 is out, this version modifies the Scoreboard.
So I noticed that the Scoreboard appears, but without the lines...

The following exception occurs:
`java.lang.RuntimeException: Unable to update scoreboard lines` [updateLines()](https://github.com/MrMicky-FR/FastBoard/blob/a9b5019b87c2c4f4c837ea4c4357ab506db99efd/src/main/java/fr/mrmicky/fastboard/FastBoardBase.java#L315)

by calling the [sendScorePacket()](https://github.com/MrMicky-FR/FastBoard/blob/a9b5019b87c2c4f4c837ea4c4357ab506db99efd/src/main/java/fr/mrmicky/fastboard/FastBoardBase.java#L472)

Cause of exception:
`Caused by: java.lang.IllegalAccessException: Can not set final java.lang.String field net.minecraft.network.protocol.game.PacketPlayOutScoreboardScore.a to java.lang.String`

The method #sendScorePacket() will invoke the PACKET_SB_SCORE by allocating the necessary fields.
Except that the PacketPlayOutScoreboardScore has changed: it's no longer a class, but a record.

I didn't see any other changes, but regarding issue #41, when an Objective is created, a new parameter is present.

Honestly, this issue is here to point out that the library doesn't work (at least for me) in version 1.20.3, after which I didn't really investigate any further.

Hopefully it was useful and not the other way around ;)

--- a/src/main/java/fr/mrmicky/fastboard/FastBoardBase.java
+++ b/src/main/java/fr/mrmicky/fastboard/FastBoardBase.java
@@ -70,11 +70,12 @@ public abstract class FastBoardBase<T> {
     // Scoreboard packets
     private static final FastReflection.PacketConstructor PACKET_SB_OBJ;
     private static final FastReflection.PacketConstructor PACKET_SB_DISPLAY_OBJ;
-    private static final FastReflection.PacketConstructor PACKET_SB_SCORE;
+    private static final MethodHandle PACKET_SB_SCORE;
     private static final FastReflection.PacketConstructor PACKET_SB_TEAM;
     private static final FastReflection.PacketConstructor PACKET_SB_SERIALIZABLE_TEAM;
     // Scoreboard enums
     private static final Class<?> DISPLAY_SLOT_TYPE;
+    private static final Class<?> NUMBER_FORMAT_INTERFACE;
     private static final Class<?> ENUM_SB_HEALTH_DISPLAY;
     private static final Class<?> ENUM_SB_ACTION;
     private static final Object SIDEBAR_DISPLAY_SLOT;
@@ -116,6 +117,12 @@ public abstract class FastBoardBase<T> {
                     )
                     .filter(m -> m.getParameterCount() == 1 && m.getParameterTypes()[0] == packetClass)
                     .findFirst().orElseThrow(NoSuchMethodException::new);
+            Class<?> numberFormat = null;
+            try {
+                numberFormat = FastReflection.nmsClass("network.chat.numbers", "NumberFormat");
+            } catch (ClassNotFoundException ignored) { // 1.20.2-
+            }
+            NUMBER_FORMAT_INTERFACE = numberFormat;
 
             Optional<Class<?>> displaySlotEnum = FastReflection.nmsOptionalClass("world.scores", "DisplaySlot");
             CHAT_COMPONENT_CLASS = FastReflection.nmsClass("network.chat", "IChatBaseComponent");
@@ -128,7 +135,18 @@ public abstract class FastBoardBase<T> {
             SEND_PACKET = lookup.unreflect(sendPacketMethod);
             PACKET_SB_OBJ = FastReflection.findPacketConstructor(packetSbObjClass, lookup);
             PACKET_SB_DISPLAY_OBJ = FastReflection.findPacketConstructor(packetSbDisplayObjClass, lookup);
-            PACKET_SB_SCORE = FastReflection.findPacketConstructor(packetSbScoreClass, lookup);
+            MethodHandle packetSbScore;
+            if (VersionType.V1_17.isHigherOrEqual()) {
+                if (numberFormat != null) { // 1.20.3+
+                    packetSbScore = lookup.findConstructor(packetSbScoreClass, MethodType.methodType(void.class, String.class, String.class, int.class, CHAT_COMPONENT_CLASS, numberFormat));
+                } else { // 1.20.2-
+                    Class<?> enumSbAction = FastReflection.nmsClass("server", "ScoreboardServer$Action");
+                    packetSbScore = lookup.findConstructor(packetSbScoreClass, MethodType.methodType(void.class, enumSbAction, String.class, String.class, int.class));
+                }
+            } else {
+                packetSbScore = lookup.findConstructor(packetSbScoreClass, MethodType.methodType(void.class));
+            }
+            PACKET_SB_SCORE = packetSbScore;
             PACKET_SB_TEAM = FastReflection.findPacketConstructor(packetSbTeamClass, lookup);
             PACKET_SB_SERIALIZABLE_TEAM = sbTeamClass == null ? null : FastReflection.findPacketConstructor(sbTeamClass, lookup);
 
@@ -470,13 +488,24 @@ public abstract class FastBoardBase<T> {
     }
 
     protected void sendScorePacket(int score, ScoreboardAction action) throws Throwable {
+        Object enumAction = action == ScoreboardAction.REMOVE
+                ? ENUM_SB_ACTION_REMOVE : ENUM_SB_ACTION_CHANGE;
+        if (VersionType.V1_17.isHigherOrEqual()) {
+            Object packet;
+            if (NUMBER_FORMAT_INTERFACE != null) { // 1.20.3+
+                packet = PACKET_SB_SCORE.invoke(COLOR_CODES[score], this.id, score, null, null);
+            } else {
+                packet = PACKET_SB_SCORE.invoke(enumAction, this.id, COLOR_CODES[score], score);
+            }
+            sendPacket(packet);
+            return;
+        }
+
         Object packet = PACKET_SB_SCORE.invoke();
 
         setField(packet, String.class, COLOR_CODES[score], 0); // Player Name
 
         if (VersionType.V1_8.isHigherOrEqual()) {
-            Object enumAction = action == ScoreboardAction.REMOVE
-                    ? ENUM_SB_ACTION_REMOVE : ENUM_SB_ACTION_CHANGE;
             setField(packet, ENUM_SB_ACTION, enumAction);
         } else {
             setField(packet, int.class, action.ordinal(), 1); // Action

--- a/src/main/java/fr/mrmicky/fastboard/FastBoardBase.java
+++ b/src/main/java/fr/mrmicky/fastboard/FastBoardBase.java
@@ -70,13 +70,12 @@ public abstract class FastBoardBase<T> {
     // Scoreboard packets
     private static final FastReflection.PacketConstructor PACKET_SB_OBJ;
     private static final FastReflection.PacketConstructor PACKET_SB_DISPLAY_OBJ;
-    private static final MethodHandle PACKET_SB_SET_SCORE;
-    private static final MethodHandle PACKET_SB_RESET_SCORE;
     private static final FastReflection.PacketConstructor PACKET_SB_TEAM;
     private static final FastReflection.PacketConstructor PACKET_SB_SERIALIZABLE_TEAM;
+    private static final MethodHandle PACKET_SB_SET_SCORE;
+    private static final MethodHandle PACKET_SB_RESET_SCORE;
     // Scoreboard enums
     private static final Class<?> DISPLAY_SLOT_TYPE;
-    private static final Class<?> NUMBER_FORMAT_INTERFACE;
     private static final Class<?> ENUM_SB_HEALTH_DISPLAY;
     private static final Class<?> ENUM_SB_ACTION;
     private static final Object SIDEBAR_DISPLAY_SLOT;
@@ -118,13 +117,6 @@ public abstract class FastBoardBase<T> {
                     )
                     .filter(m -> m.getParameterCount() == 1 && m.getParameterTypes()[0] == packetClass)
                     .findFirst().orElseThrow(NoSuchMethodException::new);
-            Class<?> numberFormat = null;
-            try {
-                numberFormat = FastReflection.nmsClass("network.chat.numbers", "NumberFormat");
-            } catch (ClassNotFoundException ignored) { // 1.20.2-
-            }
-            NUMBER_FORMAT_INTERFACE = numberFormat;
-
             Optional<Class<?>> displaySlotEnum = FastReflection.nmsOptionalClass("world.scores", "DisplaySlot");
             CHAT_COMPONENT_CLASS = FastReflection.nmsClass("network.chat", "IChatBaseComponent");
             CHAT_FORMAT_ENUM = FastReflection.nmsClass(null, "EnumChatFormat");
@@ -136,20 +128,25 @@ public abstract class FastBoardBase<T> {
             SEND_PACKET = lookup.unreflect(sendPacketMethod);
             PACKET_SB_OBJ = FastReflection.findPacketConstructor(packetSbObjClass, lookup);
             PACKET_SB_DISPLAY_OBJ = FastReflection.findPacketConstructor(packetSbDisplayObjClass, lookup);
+
+            Optional<Class<?>> numberFormat = FastReflection.nmsOptionalClass("network.chat.numbers", "NumberFormat");
             MethodHandle packetSbSetScore;
             MethodHandle packetSbResetScore = null;
-            if (VersionType.V1_17.isHigherOrEqual()) {
-                if (numberFormat != null) { // 1.20.3+
-                    packetSbSetScore = lookup.findConstructor(packetSbScoreClass, MethodType.methodType(void.class, String.class, String.class, int.class, CHAT_COMPONENT_CLASS, numberFormat));
-                    Class<?> resetScoreClass = FastReflection.nmsClass(gameProtocolPackage, "ClientboundResetScorePacket");
-                    packetSbResetScore = lookup.findConstructor(resetScoreClass, MethodType.methodType(void.class, String.class, String.class));
-                } else { // 1.20.2-
-                    Class<?> enumSbAction = FastReflection.nmsClass("server", "ScoreboardServer$Action");
-                    packetSbSetScore = lookup.findConstructor(packetSbScoreClass, MethodType.methodType(void.class, enumSbAction, String.class, String.class, int.class));
-                }
+
+            if (numberFormat.isPresent()) { // 1.20.3
+                Class<?> resetScoreClass = FastReflection.nmsClass(gameProtocolPackage, "ClientboundResetScorePacket");
+                MethodType setScoreType = MethodType.methodType(void.class, String.class, String.class, int.class, CHAT_COMPONENT_CLASS, numberFormat.get());
+                MethodType removeScoreType = MethodType.methodType(void.class, String.class, String.class);
+                packetSbSetScore = lookup.findConstructor(packetSbScoreClass, setScoreType);
+                packetSbResetScore = lookup.findConstructor(resetScoreClass, removeScoreType);
+            } else if (VersionType.V1_17.isHigherOrEqual()) {
+                Class<?> enumSbAction = FastReflection.nmsClass("server", "ScoreboardServer$Action");
+                MethodType scoreType = MethodType.methodType(void.class, enumSbAction, String.class, String.class, int.class);
+                packetSbSetScore = lookup.findConstructor(packetSbScoreClass, scoreType);
             } else {
                 packetSbSetScore = lookup.findConstructor(packetSbScoreClass, MethodType.methodType(void.class));
             }
+
             PACKET_SB_SET_SCORE = packetSbSetScore;
             PACKET_SB_RESET_SCORE = packetSbResetScore;
             PACKET_SB_TEAM = FastReflection.findPacketConstructor(packetSbTeamClass, lookup);
@@ -493,18 +490,8 @@ public abstract class FastBoardBase<T> {
     }
 
     protected void sendScorePacket(int score, ScoreboardAction action) throws Throwable {
-        Object enumAction = action == ScoreboardAction.REMOVE
-                ? ENUM_SB_ACTION_REMOVE : ENUM_SB_ACTION_CHANGE;
         if (VersionType.V1_17.isHigherOrEqual()) {
-            Object packet;
-            if (action == ScoreboardAction.CHANGE && NUMBER_FORMAT_INTERFACE != null) { // 1.20.3+
-                packet = PACKET_SB_SET_SCORE.invoke(COLOR_CODES[score], this.id, score, null, null);
-            } else if (PACKET_SB_RESET_SCORE != null) { // 1.20.3+
-                packet = PACKET_SB_RESET_SCORE.invoke(COLOR_CODES[score], this.id);
-            } else {
-                packet = PACKET_SB_SET_SCORE.invoke(enumAction, this.id, COLOR_CODES[score], score);
-            }
-            sendPacket(packet);
+            sendModernScorePacket(score, action);
             return;
         }
 
@@ -513,6 +500,8 @@ public abstract class FastBoardBase<T> {
         setField(packet, String.class, COLOR_CODES[score], 0); // Player Name
 
         if (VersionType.V1_8.isHigherOrEqual()) {
+            Object enumAction = action == ScoreboardAction.REMOVE
+                    ? ENUM_SB_ACTION_REMOVE : ENUM_SB_ACTION_CHANGE;
             setField(packet, ENUM_SB_ACTION, enumAction);
         } else {
             setField(packet, int.class, action.ordinal(), 1); // Action
@@ -524,6 +513,24 @@ public abstract class FastBoardBase<T> {
         }
 
         sendPacket(packet);
+    }
+
+    private void sendModernScorePacket(int score, ScoreboardAction action) throws Throwable {
+        String objName = COLOR_CODES[score];
+        Object enumAction = action == ScoreboardAction.REMOVE
+                ? ENUM_SB_ACTION_REMOVE : ENUM_SB_ACTION_CHANGE;
+
+        if (PACKET_SB_RESET_SCORE == null) { // Pre 1.20.3
+            sendPacket(PACKET_SB_SET_SCORE.invoke(enumAction, this.id, objName, score));
+            return;
+        }
+
+        if (action == ScoreboardAction.REMOVE) {
+            sendPacket(PACKET_SB_RESET_SCORE.invoke(objName, this.id));
+            return;
+        }
+
+        sendPacket(PACKET_SB_SET_SCORE.invoke(objName, this.id, score, null, null));
     }
 
     protected void sendTeamPacket(int score, TeamMode mode) throws Throwable {

--- a/src/main/java/fr/mrmicky/fastboard/FastBoardBase.java
+++ b/src/main/java/fr/mrmicky/fastboard/FastBoardBase.java
@@ -70,7 +70,8 @@ public abstract class FastBoardBase<T> {
     // Scoreboard packets
     private static final FastReflection.PacketConstructor PACKET_SB_OBJ;
     private static final FastReflection.PacketConstructor PACKET_SB_DISPLAY_OBJ;
-    private static final MethodHandle PACKET_SB_SCORE;
+    private static final MethodHandle PACKET_SB_SET_SCORE;
+    private static final MethodHandle PACKET_SB_RESET_SCORE;
     private static final FastReflection.PacketConstructor PACKET_SB_TEAM;
     private static final FastReflection.PacketConstructor PACKET_SB_SERIALIZABLE_TEAM;
     // Scoreboard enums
@@ -135,18 +136,22 @@ public abstract class FastBoardBase<T> {
             SEND_PACKET = lookup.unreflect(sendPacketMethod);
             PACKET_SB_OBJ = FastReflection.findPacketConstructor(packetSbObjClass, lookup);
             PACKET_SB_DISPLAY_OBJ = FastReflection.findPacketConstructor(packetSbDisplayObjClass, lookup);
-            MethodHandle packetSbScore;
+            MethodHandle packetSbSetScore;
+            MethodHandle packetSbResetScore = null;
             if (VersionType.V1_17.isHigherOrEqual()) {
                 if (numberFormat != null) { // 1.20.3+
-                    packetSbScore = lookup.findConstructor(packetSbScoreClass, MethodType.methodType(void.class, String.class, String.class, int.class, CHAT_COMPONENT_CLASS, numberFormat));
+                    packetSbSetScore = lookup.findConstructor(packetSbScoreClass, MethodType.methodType(void.class, String.class, String.class, int.class, CHAT_COMPONENT_CLASS, numberFormat));
+                    Class<?> resetScoreClass = FastReflection.nmsClass(gameProtocolPackage, "ClientboundResetScorePacket");
+                    packetSbResetScore = lookup.findConstructor(resetScoreClass, MethodType.methodType(void.class, String.class, String.class));
                 } else { // 1.20.2-
                     Class<?> enumSbAction = FastReflection.nmsClass("server", "ScoreboardServer$Action");
-                    packetSbScore = lookup.findConstructor(packetSbScoreClass, MethodType.methodType(void.class, enumSbAction, String.class, String.class, int.class));
+                    packetSbSetScore = lookup.findConstructor(packetSbScoreClass, MethodType.methodType(void.class, enumSbAction, String.class, String.class, int.class));
                 }
             } else {
-                packetSbScore = lookup.findConstructor(packetSbScoreClass, MethodType.methodType(void.class));
+                packetSbSetScore = lookup.findConstructor(packetSbScoreClass, MethodType.methodType(void.class));
             }
-            PACKET_SB_SCORE = packetSbScore;
+            PACKET_SB_SET_SCORE = packetSbSetScore;
+            PACKET_SB_RESET_SCORE = packetSbResetScore;
             PACKET_SB_TEAM = FastReflection.findPacketConstructor(packetSbTeamClass, lookup);
             PACKET_SB_SERIALIZABLE_TEAM = sbTeamClass == null ? null : FastReflection.findPacketConstructor(sbTeamClass, lookup);
 
@@ -492,16 +497,18 @@ public abstract class FastBoardBase<T> {
                 ? ENUM_SB_ACTION_REMOVE : ENUM_SB_ACTION_CHANGE;
         if (VersionType.V1_17.isHigherOrEqual()) {
             Object packet;
-            if (NUMBER_FORMAT_INTERFACE != null) { // 1.20.3+
-                packet = PACKET_SB_SCORE.invoke(COLOR_CODES[score], this.id, score, null, null);
+            if (action == ScoreboardAction.CHANGE && NUMBER_FORMAT_INTERFACE != null) { // 1.20.3+
+                packet = PACKET_SB_SET_SCORE.invoke(COLOR_CODES[score], this.id, score, null, null);
+            } else if (PACKET_SB_RESET_SCORE != null) { // 1.20.3+
+                packet = PACKET_SB_RESET_SCORE.invoke(COLOR_CODES[score], this.id);
             } else {
-                packet = PACKET_SB_SCORE.invoke(enumAction, this.id, COLOR_CODES[score], score);
+                packet = PACKET_SB_SET_SCORE.invoke(enumAction, this.id, COLOR_CODES[score], score);
             }
             sendPacket(packet);
             return;
         }
 
-        Object packet = PACKET_SB_SCORE.invoke();
+        Object packet = PACKET_SB_SET_SCORE.invoke();
 
         setField(packet, String.class, COLOR_CODES[score], 0); // Player Name
 

